logstash-core/src/main/java/org/logstash/instrument/witness/PluginsWitness.java

```diff
@@ -0,0 +1,99 @@
+package org.logstash.instrument.witness;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A Witness for the set of plugins.
+ */
+public class PluginsWitness{
+
+    private final Map<String, PluginWitness> inputs;
+    private final Map<String, PluginWitness> outputs;
+    private final Map<String, PluginWitness> filters;
+    private final Forgetter forgetter;
+
+    /**
+     * Constructor.
+     */
+    public PluginsWitness() {
+
+        this.inputs = new HashMap<>();
+        this.outputs = new HashMap<>();
+        this.filters = new HashMap<>();
+        this.forgetter = new Forgetter(this);
+    }
+
+    /**
+     * Gets the {@link PluginWitness} for the given id, creates the associated {@link PluginWitness} if needed
+     * @param id the id of the input
+     * @return the associated {@link PluginWitness} (for method chaining)
+     */
+    public PluginWitness inputs(String id) {
+        return getPlugin(inputs, id);
+    }
+
+    /**
+     * Gets the {@link PluginWitness} for the given id, creates the associated {@link PluginWitness} if needed
+     * @param id the id of the output
+     * @return the associated {@link PluginWitness} (for method chaining)
+     */
+    public PluginWitness outputs(String id) {
+        return getPlugin(outputs, id);
+    }
+
+    /**
+     * Gets the {@link PluginWitness} for the given id, creates the associated {@link PluginWitness} if needed
+     * @param id the id of the filter
+     * @return the associated {@link PluginWitness} (for method chaining)
+     */
+    public PluginWitness filters(String id) {
+        return getPlugin(filters, id);
+    }
+
+    /**
+     * Gets the {@link Forgetter} to help reset underlying metrics
+     * @return The associated {@link Forgetter}
+     */
+    public Forgetter forget() {
+        return forgetter;
+    }
+
+    /**
+     * Gets or creates the {@link PluginWitness}
+     * @param plugin the map of the plugin type.
+     * @param id the id of the plugin
+     * @return existing or new {@link PluginWitness}
+     */
+    private PluginWitness getPlugin(Map<String, PluginWitness> plugin, String id) {
+        if (plugin.containsKey(id)) {
+            return plugin.get(id);
+        } else {
+            PluginWitness pluginWitness = new PluginWitness(id);
```

Comment:
Also, one of our design goals is moving away from `lazy` metrics appearing. I think it'd be good to ensure all plugins without any laziness if possible, and to enforce that in these metrics classes as much as possible.